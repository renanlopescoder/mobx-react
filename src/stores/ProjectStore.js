import { types } from "mobx-state-tree";

import Project from './models/Project';
import ProjectActions from './actions/ProjectActions';
import ProjectViews from './views/ProjectViews';

const ProjectStore = types.model('ProjectStore', {
  loading: types.maybe(types.boolean),
  projects: types.map(Project),
})
.views(ProjectViews)
.actions(ProjectActions);


export default ProjectStore;
