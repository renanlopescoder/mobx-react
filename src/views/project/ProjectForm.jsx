import React from 'react'
import { inject, observer } from 'mobx-react'
import { Row, Col, Grid } from 'react-flexbox-grid'
import {
  Input,
  CardTitle,
  CardText,
  DatePicker,
  Dropdown,
  Autocomplete,
  Button,
  Snackbar,
} from 'react-toolbox'
import { Link } from 'react-router-dom'

import { Card } from '../../components'

@inject('projectStore')
@observer
class ProjectForm extends React.Component {
  constructor(props) {
    super(props);
    this.state ={
      author: '',
      demoLink: '',
      description: '',
      authorLink: '',
      githubLink: '',
      project: '',
      status: '',
      technologies: '',
      activeSnackbar: false,
    };
  };

  handleState = (field, value) => {
    const newState = this.state;
    newState[field] = value;
    this.setState(newState);
  };

  handleSnackbarClick = () => {
    this.setState({ activeSnackbar: false });
  };


  validForm() {
    const { state } = this;
    return (
      state.author !== '' &&
      state.demoLink !== '' &&
      state.description !== '' &&
      state.authorLink !== '' &&
      state.githubLink !== '' &&
      state.project !== '' &&
      state.status !== '' &&
      state.technologies !== ''
    );
  };

  validateForm = () => {
    const valid = this.validForm();
    this.setState({ activeSnackbar: !valid });
    return valid;
  };

  register() {
    if (this.validateForm()) {
      this.props.projectStore.createProject(this.state).then(
        created => this.props.history.push('/projects')
      )
    }
  }

  snackbarJSX = () => (
    <Snackbar
      label='Todos os campos são obrigatórios'
      onClick={this.handleSnackbarClick}
      active={this.state.activeSnackbar}
      type='warning'
      action='Close'
    />
  );

  inputJSX = (value, label, type, state) => (
    <Input
      value={value}
      label={label}
      type={type}
      onChange={change => this.handleState(state, change)}
      required
    />
  );

  render() {
    return (
      <Grid fluid>
        <Card>
          <CardTitle title='Project' subtitle='New Project' />
          <CardText>
            <Row>
              <Col md={2}>
                {this.inputJSX(this.state.project, 'Name', 'text', 'project')}
              </Col>
              <Col md={4}>
                {this.inputJSX(this.state.description, 'Description', 'text', 'description')}
              </Col>
              <Col md={2}>
                {this.inputJSX(this.state.author, 'Author', 'text', 'author')}
              </Col>
              <Col md={4}>
                {this.inputJSX(this.state.technologies, 'Technologies', 'text', 'technologies')}
              </Col>
            </Row>
            <Row>
              <Col md={2}>
                {this.inputJSX(this.state.status, 'Status', 'text', 'status')}
              </Col>
              <Col md={4}>
                {this.inputJSX(this.state.demoLink, 'DEMO Link', 'text', 'demoLink')}
              </Col>
              <Col md={3}>
                {this.inputJSX(this.state.githubLink, 'GutHub Link', 'text', 'githubLink')}
              </Col>
              <Col md={3}>
                {this.inputJSX(this.state.authorLink, 'Author Profile', 'text', 'authorLink')}
              </Col>
            </Row>
            <Row>
              <Col mdOffset={5} xs={12} md={2}>
                <Button raised primary label='Cadastrar' onClick={() => this.register()} />
              </Col>
            </Row>
          </CardText>
          {this.snackbarJSX()}
        </Card>
      </Grid>
    )
  }
}

export default ProjectForm;
