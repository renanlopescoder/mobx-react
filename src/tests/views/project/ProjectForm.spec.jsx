import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'

import ProjectStore from '../../../stores/ProjectStore'
import ProjectForm from '../../../views/project/ProjectForm'

describe('view/<ProjectForm />', () => {
  const projectStore = ProjectStore.create()
  it('should render the question header', () => {
    const component = renderer.create(
      <MemoryRouter>
        <ProjectForm projectStore={projectStore} />
      </MemoryRouter>,
    )
    const json = component.toJSON()
    expect(json).toMatchSnapshot()
  })
})
