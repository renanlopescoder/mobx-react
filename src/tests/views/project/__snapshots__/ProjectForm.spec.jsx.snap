// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view/<ProjectForm /> should render the question header 1`] = `
<div
  className="container-fluid"
>
  <div
    className=""
    data-react-toolbox="card"
  >
    <div
      className="undefined"
    >
      <div>
        <h5
          className={undefined}
        >
          Project
        </h5>
        <p
          className={undefined}
        >
          New Project
        </p>
      </div>
    </div>
    <div
      className=""
    >
      <div
        className="row"
      >
        <div
          className="col-md-2"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              Name
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-md-4"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              Description
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-md-2"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              Author
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-md-4"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              Technologies
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-2"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              Status
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-md-4"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              DEMO Link
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-md-3"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              GutHub Link
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-md-3"
        >
          <div
            className=""
            data-react-toolbox="input"
          >
            <input
              className=""
              defaultValue={undefined}
              disabled={false}
              maxLength={undefined}
              name={undefined}
              onChange={[Function]}
              onKeyPress={undefined}
              required={true}
              role="input"
              type="text"
              value=""
            />
            <span
              className={undefined}
            />
            <label
              className=""
            >
              Author Profile
              <span
                className={undefined}
              >
                 * 
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-offset-5 col-xs-12 col-md-2"
        >
          <button
            className=""
            data-react-toolbox="button"
            disabled={false}
            href={undefined}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchStart={[Function]}
            type="button"
          >
            Cadastrar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
